src/serialization/compact/CompactStreamSerializer.ts

```diff
@@ -0,0 +1,179 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @ignore *//** */
+
+import {CompactSerializer} from './CompactSerializer';
+import {Schema} from './Schema';
+import {DefaultCompactReader} from './DefaultCompactReader';
+import {SchemaService} from './SchemaService';
+import {DefaultCompactWriter} from './DefaultCompactWriter';
+import {FieldOperations} from '../generic_record/FieldOperations';
+import {HazelcastSerializationError, SchemaNotFoundError, SchemaNotReplicatedError} from '../../core';
+import {ObjectDataInput, ObjectDataOutput, PositionalObjectDataOutput} from '../ObjectData';
+import {CompactGenericRecordImpl} from '../generic_record/CompactGenericRecord';
+import {SchemaWriter} from './SchemaWriter';
+
+/**
+ * Serializer for compact serializable objects.
+ *
+ * This serializer is used for compact serializable objects that are registered using serialization config
+ * and Compact generic records ({@link CompactGenericRecordImpl}).
+ *
+ * @internal
+ */
+export class CompactStreamSerializer {
+
+    id = -55;
+    /**
+     * Users' serializer config for classes are stored here. Used to determine if a class is compact serializable.
+     * Also used to get serializer of an object while serializing.
+     */
+    private static readonly classToSerializersMap:
+     Map<new (...args: any[]) => any, CompactSerializer<new (...args: any[]) => any>> = new Map();
+    /**
+     * Used to cache created schema of an object after initial serialization. If an object has schema,
+     * no need to create schema again and put to schema service.
+     */
+    private readonly classToSchemaMap : Map<new (...args: any[]) => any, Schema>;
+    /**
+     * A table from typeName to serializer map. Serialized compact data include a type name. Deserializer of compact data is
+     * determined using this table.
+     */
+    private readonly typeNameToSerializersMap: Map<string, CompactSerializer<new (...args: any[]) => any>>;
+
+    constructor(
+        private readonly schemaService: SchemaService
+    ) {
+        this.classToSchemaMap = new Map<new (...args: any[]) => any, Schema>();
+        this.typeNameToSerializersMap = new Map<string, CompactSerializer<new (...args: any[]) => any>>();
+    }
+
+    getOrReadSchema(input: ObjectDataInput): Schema {
+        const schemaId = input.readLong();
+        const schema = this.schemaService.get(schemaId);
+        if (schema !== null) {
+            return schema;
+        }
+
+        throw new SchemaNotFoundError(`The schema can not be found with id ${schemaId}`, schemaId);
```

Comment:
I think the error is not internal since we annotate some of our APIs with it. (Lazy deserialization constructs' get methods). Also user can directly receive this error unlike SchemaNotReplicatedError. If we remove the message it won't be very helpful to users since they cannot see the schema id without printing it.